client/consensus/grandpa/authorities.go

```diff
@@ -0,0 +1,908 @@
+	// The new authorities and weights to apply.
+	nextAuthorities []Authority
+	// How deep in the chain the announcing block must be
+	// before the hashNumber is applied.
+	delay N
+	// The announcing block's height.
+	canonHeight N
+	// The announcing block's hash.
+	canonHash H
+	// The delay kind.
+	delayKind delayKind
+}
+
+// EffectiveNumber Returns the effective number this hashNumber will be applied at.
+func (pc *PendingChange[H, N]) EffectiveNumber() N {
+	return pc.canonHeight + pc.delay
+}
+
+// AuthoritySetChanges Tracks historical authority set changes. We store the block numbers for the last block
+// of each authority set, once they have been finalised. These blocks are guaranteed to
+// have a justification unless they were triggered by a forced hashNumber.
+type AuthoritySetChanges[N constraints.Unsigned] []setIDNumber[N]
+
+// append an setIDNumber to AuthoritySetChanges
+func (asc *AuthoritySetChanges[N]) append(setID uint64, blockNumber N) {
+	*asc = append(*asc, setIDNumber[N]{
+		setID:       setID,
+		blockNumber: blockNumber,
+	})
+}
+
+type authoritySetChangeID scale.VaryingDataType
+
+// Set will set a VaryingDataTypeValue using the underlying VaryingDataType
+func (asc *authoritySetChangeID) Set(val scale.VaryingDataTypeValue) (err error) {
+	vdt := scale.VaryingDataType(*asc)
+	err = vdt.Set(val)
+	if err != nil {
+		return
+	}
+	*asc = authoritySetChangeID(vdt)
+	return
+}
+
+// Value will return value from underying VaryingDataType
+func (asc *authoritySetChangeID) Value() (val scale.VaryingDataTypeValue, err error) {
+	vdt := scale.VaryingDataType(*asc)
+	return vdt.Value()
+}
+
+func newAuthoritySetChangeID[N constraints.Unsigned]() authoritySetChangeID {
+	vdt, err := scale.NewVaryingDataType(latest{}, set[N]{}, unknown{})
+	if err != nil {
+		panic(err)
+	}
+	return authoritySetChangeID(vdt)
+}
+
+type latest struct{}
+
+func (latest) Index() uint {
+	return 0
+}
+
+type set[N constraints.Unsigned] struct {
+	inner setIDNumber[N]
+}
+
+func (set[N]) Index() uint {
+	return 1
+}
+
+type unknown struct{}
+
+func (unknown) Index() uint {
+	return 2
+}
+
+// Three states that can be returned: Latest, Set (tuple), Unknown
+func (asc *AuthoritySetChanges[N]) getSetID(blockNumber N) (authSetChangeID authoritySetChangeID, err error) {
+	if asc == nil {
+		return authSetChangeID, fmt.Errorf("getSetID: authSetChanges is nil")
+	}
+	authSetChangeID = newAuthoritySetChangeID[N]()
+	authSet := *asc
+	last := authSet[len(authSet)-1]
+	if last.blockNumber < blockNumber {
+		err = authSetChangeID.Set(latest{})
+		if err != nil {
+			return authSetChangeID, err
+		}
+		return authSetChangeID, nil
+	}
+
+	idx, _ := slices.BinarySearchFunc(
+		authSet,
+		blockNumber,
+		func(a setIDNumber[N], b N) int {
+			switch {
+			case a.blockNumber == b:
+				return 0
+			case a.blockNumber > b:
+				return 1
+			case a.blockNumber < b:
+				return -1
+			default:
+				panic("huh?")
+			}
```

Comment:
looks like all possible comparisons are covered so I think you can remove the `default` case

Reply:
All cases of covered but a return is still needed since it's not literally every possible case. I have kept as is and modified the panic